src/hooks/useReportIDs.tsx

```diff
@@ -0,0 +1,135 @@
+});
+
+function WithReportIDsContextProvider({
+    children,
+    chatReports,
+    betas,
+    policies,
+    allReportActions,
+    transactionViolations,
+    policyMembers,
+    priorityMode,
+    /**
+     * Only required to make unit tests work, since we
+     * explicitly pass the currentReportID in LHNTestUtils
+     * to SidebarLinksData, so this context doesn't have an
+     * access to currentReportID in that case.
+     *
+     * This is a workaround to have currentReportID available in testing environment.
+     */
+    currentReportIDForTests,
+}: WithReportIDsContextProviderProps) {
+    const currentReportIDValue = useCurrentReportID();
+    const derivedCurrentReportID = currentReportIDForTests ?? currentReportIDValue?.currentReportID;
+    const {activeWorkspaceID} = useActiveWorkspace();
+
+    const policyMemberAccountIDs = useMemo(() => getPolicyMembersByIdWithoutCurrentUser(policyMembers, activeWorkspaceID, getCurrentUserAccountID()), [activeWorkspaceID, policyMembers]);
+
+    const getOrderedReportIDs = useCallback(
+        (currentReportID?: string) =>
+            SidebarUtils.getOrderedReportIDs(
+                currentReportID ?? null,
+                chatReports,
+                betas ?? [],
+                policies,
+                priorityMode,
+                allReportActions,
+                transactionViolations,
+                activeWorkspaceID,
+                policyMemberAccountIDs,
+            ),
+        [chatReports, betas, policies, priorityMode, allReportActions, transactionViolations, activeWorkspaceID, policyMemberAccountIDs],
+    );
+
+    const orderedReportIDs = useMemo(() => getOrderedReportIDs(), [getOrderedReportIDs]);
+
+    // We need to make sure the current report is in the list of reports, but we do not want
+    // to have to re-generate the list every time the currentReportID changes. To do that
+    // we first generate the list as if there was no current report, then we check if
+    // the current report is missing from the list, which should very rarely happen. In this
+    // case we re-generate the list a 2nd time with the current report included.
+    const orderedReportIDsWithCurrentReport = useMemo(() => {
+        if (derivedCurrentReportID && !orderedReportIDs.includes(derivedCurrentReportID)) {
+            return getOrderedReportIDs(derivedCurrentReportID);
+        }
+        return orderedReportIDs;
+    }, [getOrderedReportIDs, derivedCurrentReportID, orderedReportIDs]);
+
+    const contextValue = useMemo(
+        () => ({
+            orderedReportIDs: orderedReportIDsWithCurrentReport,
+        }),
+        [orderedReportIDsWithCurrentReport],
+    );
+
+    return <ReportIDsContext.Provider value={contextValue}>{children}</ReportIDsContext.Provider>;
+}
+
+const ReportIDsContextProvider = withOnyx<WithReportIDsContextProviderProps, OnyxProps>({
+    chatReports: {
+        key: ONYXKEYS.COLLECTION.REPORT,
+        initialValue: {},
```

Comment:
in the original code inside `SidebarlinksData`, the `withOnyx` uses a selector to narrow down the properties from the report object to be used. Should we do the same here or that is no longer necessary because of the new optimization?

Also, more high-level question. Is there a reason you moved `withOnyx` to `useReportIds` from `SidebarlinksData`?

Reply:
>Should we do the same here or that is no longer necessary because of the new optimization?

I can't say that it's no longer necessary but I also didn't observe any performance hits without the selector in this PR. I tried adding the selectors but the issue is that I was facing huge TS issues which I couldn't figure out due to business logic. Last week, a PR got merged to main which migrated the `SidebarlinksData` to TS and when I pulled main, I resolved the conflicts and moved all of the selectors to the `useReportIDs` and it was all good.

But later that PR got reverted and I had to remove all of the selectors again. I guess we can either wait until that PR is merged again, then we sync this PR and add the selectors. Because if this PR gets merged first, we won't have the selectors migrated to TS.


> Is there a reason you moved withOnyx to useReportIds from SidebarlinksData?

The reason is that we don't need any Onyx data in `SidebarlinksData` anymore. Because it was only required to generate the `orderedReportIDs` which we are now doing in `useReportIDs`, so it's safe to remove `withOnyx` from `SidebarlinksData`.

Reply:
I agree with your decision regarding the use of the `selector` property.

Regarding the `withOnyx` location (i.e., where we subscribe to the Onyx data necessary for calculating ordered report IDs), I understand that it can't occur inside `SidebarlinksData`.

However, I also don't believe the Content Provider should subscribe to the Onyx data. I've added [[a comment here](https://github.com/Expensify/App/pull/36420#discussion_r1529016192)](https://github.com/Expensify/App/pull/36420#discussion_r1529016192) concerning this point.